File: backend/app/schemas/inventory.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class InboundDocSubmitRequest(BaseModel):
    actual_qty: Decimal = Field(gt=0, description="实际入库数量")
    warehouse_id: str = Field(min_length=1, max_length=64, description="仓库ID")


class OutboundDocSubmitRequest(BaseModel):
    actual_qty: Decimal = Field(gt=0, description="实际出库数量")
    warehouse_id: str = Field(min_length=1, max_length=64, description="仓库ID")


class OutboundDocWarehouseConfirmRequest(BaseModel):
    contract_id: int = Field(gt=0, description="销售合同ID")
    sales_order_id: int = Field(gt=0, description="销售订单ID")
    source_ticket_no: str = Field(
        min_length=1, max_length=64, description="仓库来源回执号"
    )
    actual_qty: Decimal = Field(gt=0, description="实际出库数量")
    warehouse_id: str = Field(min_length=1, max_length=64, description="仓库ID")


class OutboundDocManualCreateRequest(BaseModel):
    contract_id: int = Field(gt=0, description="销售合同ID")
    sales_order_id: int = Field(gt=0, description="销售订单ID")
    oil_product_id: str = Field(min_length=1, max_length=64, description="油品ID")
    manual_ref_no: str = Field(min_length=1, max_length=64, description="手工回执号")
    actual_qty: Decimal = Field(gt=0, description="实际出库数量")
    reason: str = Field(min_length=1, max_length=256, description="手工补录原因")


class InboundDocResponse(BaseModel):
    id: int
    doc_no: str
    contract_id: int
    purchase_order_id: int | None = None
    oil_product_id: str
    warehouse_id: str | None = None
    source_type: str
    actual_qty: Decimal
    status: str
    submitted_at: datetime | None = None
    created_at: datetime
    message: str


class OutboundDocResponse(BaseModel):
    id: int
    doc_no: str
    contract_id: int
    sales_order_id: int
    oil_product_id: str
    warehouse_id: str | None = None
    source_type: str
    source_ticket_no: str | None = None
    manual_ref_no: str | None = None
    actual_qty: Decimal
    status: str
    submitted_at: datetime | None = None
    created_at: datetime
    message: str


class InboundDocListItem(BaseModel):
    id: int
    doc_no: str
    contract_id: int
    purchase_order_id: int | None = None
    oil_product_id: str
    warehouse_id: str | None = None
    source_type: str
    actual_qty: Decimal
    status: str
    submitted_at: datetime | None = None
    created_at: datetime


class OutboundDocListItem(BaseModel):
    id: int
    doc_no: str
    contract_id: int
    sales_order_id: int
    oil_product_id: str
    warehouse_id: str | None = None
    source_type: str
    source_ticket_no: str | None = None
    manual_ref_no: str | None = None
    actual_qty: Decimal
    status: str
    submitted_at: datetime | None = None
    created_at: datetime


class InboundDocListResponse(BaseModel):
    items: list[InboundDocListItem]
    total: int
    message: str


class OutboundDocListResponse(BaseModel):
    items: list[OutboundDocListItem]
    total: int
    message: str

File: admin-web/src/api/inventory.ts
import {
  getDemoInboundDocDetail,
  getDemoOutboundDocDetail,
  listDemoInboundDocs,
  listDemoOutboundDocs,
} from '@/mock/inventory'

import { httpClient, reportsMode } from './http'

export interface InboundDocListItem {
  id: number
  doc_no: string
  contract_id: number
  purchase_order_id: number | null
  oil_product_id: string
  warehouse_id: string | null
  source_type: string
  actual_qty: string
  status: string
  submitted_at: string | null
  created_at: string
}

export interface OutboundDocListItem {
  id: number
  doc_no: string
  contract_id: number
  sales_order_id: number
  oil_product_id: string
  warehouse_id: string | null
  source_type: string
  source_ticket_no: string | null
  manual_ref_no: string | null
  actual_qty: string
  status: string
  submitted_at: string | null
  created_at: string
}

export interface InboundDocListResponse {
  items: InboundDocListItem[]
  total: number
  message: string
}

export interface OutboundDocListResponse {
  items: OutboundDocListItem[]
  total: number
  message: string
}

export interface InboundDocDetailResponse extends InboundDocListItem {
  message: string
}

export interface OutboundDocDetailResponse extends OutboundDocListItem {
  message: string
}

export async function fetchInboundDocs(params?: {
  status?: string
  source_type?: string
}): Promise<InboundDocListResponse> {
  if (reportsMode === 'demo') {
    return Promise.resolve(listDemoInboundDocs(params))
  }
  const { data } = await httpClient.get<InboundDocListResponse>('/inbound-docs', {
    params: {
      ...(params?.status ? { status: params.status } : {}),
      ...(params?.source_type ? { source_type: params.source_type } : {}),
      limit: 50,
    },
  })
  return data
}

export async function fetchOutboundDocs(params?: {
  status?: string
  source_type?: string
}): Promise<OutboundDocListResponse> {
  if (reportsMode === 'demo') {
    return Promise.resolve(listDemoOutboundDocs(params))
  }
  const { data } = await httpClient.get<OutboundDocListResponse>('/outbound-docs', {
    params: {
      ...(params?.status ? { status: params.status } : {}),
      ...(params?.source_type ? { source_type: params.source_type } : {}),
      limit: 50,
    },
  })
  return data
}

export async function fetchInboundDocDetail(docId: number): Promise<InboundDocDetailResponse> {
  if (reportsMode === 'demo') {
    return Promise.resolve(getDemoInboundDocDetail(docId))
  }
  const { data } = await httpClient.get<InboundDocDetailResponse>(`/inbound-docs/${docId}`)
  return data
}

export async function fetchOutboundDocDetail(docId: number): Promise<OutboundDocDetailResponse> {
  if (reportsMode === 'demo') {
    return Promise.resolve(getDemoOutboundDocDetail(docId))
  }
  const { data } = await httpClient.get<OutboundDocDetailResponse>(`/outbound-docs/${docId}`)
  return data
}
